pyscf/pbc/mp/kmp2_stagger.py

```diff
@@ -0,0 +1,456 @@
+                Lov[iki, ika] = out.reshape(-1, nocc, nvir)
+
+    log.timer_debug1("transforming DF-MP2 integrals", *cput0)
+
+    return Lov
+
+
+#   Class for staggered mesh MP2
+class KMP2_stagger(kmp2.KMP2):
+    def __init__(self, mf, frozen=None, flag_submesh=False):
+        self.cell = mf.cell
+        self._scf = mf
+        self.verbose = self.cell.verbose
+        self.stdout = self.cell.stdout
+        self.max_memory = self.cell.max_memory
+
+        #   MP2 energy
+        self.e_corr = None
+
+        #   nocc and nmo variables
+        self._nocc = None
+        self._nmo = None
+
+        #   Construction of orbitals and orbital energies on two staggered meshes
+        nks = get_monkhorst_pack_size(mf.cell, mf.kpts)  #   mesh size
+        self.flag_submesh = flag_submesh
+        if flag_submesh is True:
+            #   Choice 1: Use two staggered submeshes of mf.kpts of half size along each dimension
+            if np.any( nks % 2) :
+                log = logger.Logger(self.stdout, self.verbose)
+                log.warn(
+                    'The k-point mesh in the input mf has odd size in certain dimension. \n'
+                    'Cannot use its submeshes for staggered mesh calculation!\n'
+                    'You may set "flag_submesh" to False')
+                raise RuntimeError
+
+            nks_ov = nks // 2
+            shift = mf.kpts[ np.argmin( np.sum(mf.kpts**2, axis=1) )]   #   shift of the original k-point mesh
+            half_shift = mf.cell.get_abs_kpts( [0.5/n for n in nks_ov] )
+            kpts_vir = shift + mf.cell.make_kpts(nks_ov, with_gamma_point=True)
+            kpts_occ = kpts_vir + half_shift
+
+            self.kpts = mf.kpts
+            self.mo_energy = mf.mo_energy
+            self.mo_coeff = mf.mo_coeff
+            self.mo_occ = get_occ(mf, mo_energy_kpts=mf.mo_energy)
+
+        else:
+            #   Choice 2: Apply non-SCF calculation to compute orbitals and orbital energies on a
+            #             staggered mesh of the same size as mf.kpts.
+            if mf.cell.dimension < 3:
+                # The non-SCF calculation of orbitals/orbital energies is currently only valid
+                # for 3D systems in PySCF. For example, when using get_band with a reference HF
+                # system of mesh 1*1*k to compute orbitals and orbital energies on a 1*1*2k mesh,
+                # the obtained mo_energy and  mo_coeff are not in good agreement with those computed
+                # from an SCF-HF calculation with mesh 1*1*2k.
+                raise NotImplementedError
+
+            half_shift = mf.cell.get_abs_kpts( [0.5/n for n in nks] )
+            kpts_vir = mf.kpts
+            kpts_occ = kpts_vir + half_shift
+            kpts = np.concatenate( (kpts_occ, kpts_vir), axis=0)
+            if isinstance(mf, scf.KRHF):
+                with lib.temporary_env(mf, exxdiv='vcut_sph', with_df=df.FFTDF(mf.cell, mf.kpts)):
```

Comment:
Is using FFTDF a must here?

Reply:
For the use of "exxdiv='vcut_sph', with_df=df.FFTDF", I'm following the comment about band structure calculation in at https://pyscf.org/user/pbc/scf.html?highlight=get_band